modules/runner/src/integrationTest/java/org/apache/ignite/internal/schemasync/ItSchemaSyncAndReplicationTest.java

```diff
@@ -0,0 +1,173 @@
+     */
+    @Test
+    void laggingSchemasPreventPartitionDataReplication() throws Exception {
+        createTestTableWith3Replicas();
+
+        final int notInhibitedNodeIndex = 0;
+        transferLeadershipsTo(notInhibitedNodeIndex);
+
+        IgniteImpl nodeToInhibitMetaStorage = cluster.node(1);
+
+        WatchListenerInhibitor listenerInhibitor = WatchListenerInhibitor.metastorageEventsInhibitor(nodeToInhibitMetaStorage);
+        listenerInhibitor.startInhibit();
+
+        try {
+            CountDownLatch rejectionTriggered = rejectionDueToMetadataLagTriggered();
+
+            updateTableSchemaAt(notInhibitedNodeIndex);
+            putToTableAt(notInhibitedNodeIndex);
+
+            assertTrue(rejectionTriggered.await(10, TimeUnit.SECONDS), "Did not see rejections due to lagging metadata");
+
+            assertTrue(solePartitionIsEmpty(nodeToInhibitMetaStorage), "Something was written to the partition");
+
+            listenerInhibitor.stopInhibit();
+
+            assertTrue(
+                    waitForCondition(() -> !solePartitionIsEmpty(nodeToInhibitMetaStorage), 10_000),
+                    "Nothing was written to partition even after inhibiting was cancelled"
+            );
+        } finally {
+            listenerInhibitor.stopInhibit();
+        }
+    }
+
+    private void createTestTableWith3Replicas() throws InterruptedException {
+        String zoneSql = "create zone test_zone with partitions=1, replicas=3";
+        String sql = "create table test (key int primary key, value varchar(20))"
+                + " with primary_zone='TEST_ZONE'";
+
+        cluster.doInSession(0, session -> {
+            executeUpdate(zoneSql, session);
+            executeUpdate(sql, session);
+        });
+
+        waitForTableToStart();
+    }
+
+    private void waitForTableToStart() throws InterruptedException {
+        // TODO: IGNITE-18733 - remove this wait because when a table creation query is executed, the table must be fully ready.
+
+        BooleanSupplier tableStarted = () -> {
+            int numberOfStartedRaftNodes = cluster.runningNodes()
+                    .map(ReplicationGroupsUtils::tablePartitionIds)
+                    .mapToInt(List::size)
+                    .sum();
+            return numberOfStartedRaftNodes == 3;
+        };
+
+        assertTrue(waitForCondition(tableStarted, 10_000), "Did not see all table RAFT nodes started");
+    }
+
+    private void transferLeadershipsTo(int nodeIndex) throws InterruptedException {
+        cluster.transferLeadershipTo(nodeIndex, MetastorageGroupId.INSTANCE);
+        cluster.transferLeadershipTo(nodeIndex, cluster.solePartitionId());
+    }
+
+    private static CountDownLatch rejectionDueToMetadataLagTriggered() {
+        Logger interceptorLogger = Logger.getLogger(CheckCatalogVersionOnAppendEntries.class.getName());
+
+        CountDownLatch rejectionTriggered = new CountDownLatch(1);
```

Comment:
Both future and latch allow to wait for some event to happen, but the latch seems to be more focused on this task. On the other hand, what's the advantage of a future over a latch?

Reply:
Ease of use, instead of `assertTrue(rejectionTriggered.await(10, TimeUnit.SECONDS), "Did not see rejections due to lagging metadata");` you can use `assertThat(future, willCompleteSuccessfully());`, up to u.